File: src/DiscordBot/Services/CommandHandlingService.vb
Imports System.Reflection
Imports Discord
Imports Discord.Commands
Imports Discord.WebSocket

Namespace DiscordBot.Services
    Public Class CommandHandlingService

        Private ReadOnly _discord As DiscordSocketClient
        Private ReadOnly _commands As CommandService
        Private _provider As IServiceProvider

        Sub New(provider As IServiceProvider, discord As DiscordSocketClient, commands As CommandService)
            _discord = discord
            _commands = commands

            AddHandler _discord.MessageReceived, AddressOf MessageReceived
        End Sub

        Async Function InitializeAsync(provider As IServiceProvider) As Task
            _provider = provider
            Await _commands.AddModulesAsync(Assembly.GetEntryAssembly())
            ' Add additional initialization code here...
        End Function

        Async Function MessageReceived(rawMessage As SocketMessage) As Task
            ' Ignore system messages and messages from bots
            Dim userMessage = TryCast(rawMessage, SocketUserMessage)
            If (IsNothing(userMessage) OrElse Not (rawMessage.Source = MessageSource.User)) Then
                Return
            End If

            Dim argPos = 0
            If Not (userMessage.HasMentionPrefix(_discord.CurrentUser, argPos)) Then
                Return
            End If

            Dim context = New SocketCommandContext(_discord, userMessage)
            Dim result = Await _commands.ExecuteAsync(context, argPos, _provider)

            If (TypeOf result Is ParseResult OrElse
            TypeOf result Is PreconditionResult OrElse
            TypeOf result Is ExecuteResult OrElse
            TypeOf result Is TypeReaderResult) AndAlso Not result.IsSuccess Then
                Await context.Channel.SendMessageAsync(result.ToString())
            End If
        End Function

    End Class
End Namespace

File: src/DiscordBot/Services/CommandHandlingService.cs
using System;
using System.Reflection;
using System.Threading.Tasks;
using Discord;
using Discord.Commands;
using Discord.WebSocket;
using DiscordBot.Data;
using LiteDB;

namespace DiscordBot.Services
{
    public class CommandHandlingService
    {
        private readonly DiscordSocketClient _discord;
        private readonly CommandService _commands;
        private IServiceProvider _provider;
        private LiteDatabase _database;

        public CommandHandlingService(IServiceProvider provider, DiscordSocketClient discord, CommandService commands, LiteDatabase database)
        {
            _discord = discord;
            _commands = commands;
            _provider = provider;
            _database = database;

            _discord.MessageReceived += MessageReceived;
        }

        public async Task InitializeAsync(IServiceProvider provider)
        {
            _provider = provider;
            await _commands.AddModulesAsync(Assembly.GetEntryAssembly());
            // Add additional initialization code here...
        }

        private async Task MessageReceived(SocketMessage rawMessage)
        {
            // Ignore system messages and messages from bots
            if (!(rawMessage is SocketUserMessage message)) return;
            if (message.Source != MessageSource.User) return;

            int argPos = 0;

            // Don't require prefix for DMs 
            if (!message.HasMentionPrefix(_discord.CurrentUser, ref argPos) && !(message.Channel is SocketDMChannel)) return;
            // Uncomment below to enable a string prefix (these should only be used with private bots!)
            // if (!message.HasStringPrefix("!", ref argPos)) return;

            var context = new SocketCommandContext(_discord, message);
            var result = await _commands.ExecuteAsync(context, argPos, _provider);

            if (result.Error.HasValue &&
                result.Error.Value == CommandError.UnknownCommand)
                return;

            if (result.Error.HasValue && 
                result.Error.Value != CommandError.UnknownCommand)
                await context.Channel.SendMessageAsync(result.ToString());

            // Add points to the user for using the bot
            // Do this asynchronously, on another task, to prevent database access (and levelup notifications?) from halting the bot
            _ = UpdateLevelAsync(context);
        }

        private Task UpdateLevelAsync(SocketCommandContext context)
        {
            var users = _database.GetCollection<User>("users");
            var user = users.FindOne(u => u.Id == context.User.Id) ?? new User { Id = context.User.Id };
            ++user.Points;
            users.Upsert(user);

            // If sending a levelup notification, flag this Task as async and remove the following line
            return Task.CompletedTask;
        }
    }
}
